node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeAlerter.java

```diff
@@ -0,0 +1,442 @@
+package com.yahoo.vespa.hosted.provision.maintenance;
+
+import com.yahoo.config.provision.Flavor;
+import com.yahoo.config.provision.NodeResources;
+import com.yahoo.config.provision.NodeType;
+import com.yahoo.jdisc.Metric;
+import com.yahoo.log.LogLevel;
+import com.yahoo.vespa.hosted.provision.Node;
+import com.yahoo.vespa.hosted.provision.NodeRepository;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import com.yahoo.vespa.hosted.provision.node.Allocation;
+
+import java.util.*;
+import java.util.function.Function;
+
+/**
+ * Performs analysis on the node repository to produce metrics that can be used for alerts if the repository is in an
+ * undesirable state.
+ * These metrics include:
+ * Spare host capacity, or how many hosts the repository can stand to lose without ending up in a situation where it's
+ * unable to find a new home for orphaned tenants.
+ * Overcommitted hosts, which tracks if there are any hosts whose capacity is less than the sum of its children's.
+ *
+ * @author mgimle
+ */
+public class NodeAlerter extends Maintainer {
+
+    private final Metric metric;
+    private final NodeRepository nodeRepository;
+    private static final Logger log = Logger.getLogger(NodeAlerter.class.getName());
+
+    NodeAlerter(NodeRepository nodeRepository,
+                Metric metric,
+                Duration interval) {
+        super(nodeRepository, interval);
+        this.nodeRepository = nodeRepository;
+        this.metric = Objects.requireNonNull(metric);
+    }
+
+    @Override
+    protected void maintain() {
+        Optional<HostFailurePath> failurePath = worstCaseHostLossLeadingToFailure();
+        if (failurePath.isPresent()) {
+            int worstCaseHostLoss = failurePath.get().hostsCausingFailure.size();
+            metric.set("hostedVespa.spareHostCapacity", worstCaseHostLoss - 1, null);
+            metric.set("hostedVespa.overcommittedHosts", countOvercommittedHosts(), null);
+        }
+    }
+
+    /**
+     * Contains the list of hosts that, upon being removed, caused an unrecoverable state,
+     * as well as the specific host and tenant which caused it.
+     */
+    public static class HostFailurePath {
```

Comment:
I'm not a huge fan of interleaving inner classes and functions like this. I suggest moving the classes to the bottom.